net.box: fix connection's `error_reconnect` state

Serpentian opened the pull request:
Currently if we kill worker fiber of the connection, which was
initialized with `reconnect_after` option, we get `error_reconnect`
state. But reconnect obviously doesn't happen.

Let's try to solve that problem) 

Closes #7448
Closes  #7473

Gerold103 commented on test/box-luatest/gh_7448_net_box_cancelled_fiber_error_test.lua:
```diff
@@ -0,0 +1,31 @@
+    t.assert(g.conn:ping())
+    -- Explicit kill of the connection's fiber
+    pcall(function()
+        g.conn:eval([[
+            box.session.push(nil)
+        ]], {}, {on_push = function()
+            require('fiber').self():cancel()
+        end})
```
It makes sense to explain in a comment that you are using pushes to reach the worker fiber somehow. But otherwise it is not really needed. This test runs only on versions where you have `conn._fiber` available and can do `conn._fiber:cancel()`. In vshard I proposed this push-hack as a way to repeat the same on 1.10 where you don't have `._fiber` member.

Serpentian replied:
I'm already used to the fact that we test changes on all versions like in vshard. You're right, there's no need for such workaround here. Fixed

locker commented on src/box/lua/net_box.c:
```diff
@@ -2673,10 +2673,15 @@ netbox_worker_f(va_list ap)
+		/*
+		 * gh-7448: if the fiber is cancelled we should set the
+		 * NETBOX_ERROR state as no reconnection will be done anyway.
+		 */
```
I think that a net.box worker fiber, like any other system fiber, should ignore `fiber.cancel` coming from the Lua world. The only way to break a connection from Lua should be by calling `conn.stop`. To achieve that, you could mark the net.box worker fiber as non-cancellable, see `fiber_set_cancellable`, but we're [planning](https://github.com/tarantool/tarantool/issues/7166) to get rid of this fiber function. Perhaps, it's time to add `FIBER_IS_SYSTEM` flag.

Serpentian replied:
This sounds reasonable: I made a ticket for that: #7473. I talked to @filonenko-mikhail, he's all right with such feature.

However, I think this patch should be merged too as it's definitely a bug. We should not set `error_reconnect` state or wait if fiber is dead. It's not a way it used to work before and it's not quite logical.

locker replied:
The bug will be fixed by marking the fiber system. Internally, a worker fiber can be cancelled only after changing the state machine state to `closed` so it can't possibly set `error_reconnect` in this case. I don't see any point in complicating the code to fix a bug that can be fixed just by marking the net.box worker fiber system.

Serpentian replied:
Done. Let's discuss the API of this feature more precisly

Serpentian commented:
Consider this review iteration as a draft one. We should at least discuss, if we need some public Lua API for user to know that the fiber is system.  Tomorrow I will mark fibers, which are listed in [Cartridge](https://github.com/tarantool/cartridge/blob/d8d48a17e9bf4505941f7f5785fd4b6a9d809918/cartridge/hotreload.lua#L156-L170), as system. Of course, except vshard's ones.

Serpentian commented on src/lib/core/fiber.c:
```diff
@@ -557,6 +557,15 @@ fiber_set_joinable(struct fiber *fiber, bool yesno)
+void
+fiber_set_system(struct fiber *fiber, bool yesno)
+{
+	if (yesno == true)
```
Checkpatch doesn't like that. But this is done for consistency with all other `fiber_set_*`. Possibly it's needed to be patched

Serpentian commented on src/lua/fiber.c:
```diff
@@ -789,6 +791,14 @@ lbox_fiber_set_joinable(struct lua_State *L)
 }
 
+static int
+lbox_fiber_is_system(struct lua_State *L)
```
I think, that the user should at least have a way to check if the fiber cannot be cancelled, as `lbox_fiber_cancel` doesn't throw an error but just ignores cancell request if `FIBER_IS_SYSTEM` is set. 

We can also throw an error in `lbox_fiber_cancel` (it's even mentioned in docs, that this function  can throw an error if it'not permitted to cancel a fiber).  I don't really like that option, as Cartridge's (possibly, not only its) code will be broken with such change.

locker replied:
I don't see what a user could possibly do with the knowledge if a particular fiber is system. Let's not extend the public API for now and silently ignore system fibers in `fiber.cancel`. If somebody asks for a way to determine if a particular fiber is system, we'll consider adding this function or a flag to `fiber.info`.

Serpentian commented on test/box-luatest/gh_7448_net_box_cancelled_fiber_error_test.lua:
```diff
@@ -0,0 +1,31 @@
+    -- Fiber kill is performed asynchronously. Half of the
+    -- second is usually enouth to kill it. But system fiber
+    -- is not supposed to be killed anyway.
+    fiber.sleep(1)
```
There will be no need for that `fiber.sleep()` as soon as all needed fibers will be marked as system ones. I'm going to walk through `fiber.info` and try to kill system fibers by their names, after which check if they are not `dead`.

Gerold103 commented on changelogs/unreleased/gh-7473-introduce-system-fiber.md:
```diff
@@ -0,0 +1,5 @@
+
+* Introduced system fibers, which cannot be cancelled from Lua. Added C API
+  for setting fiber as system: `fiber_set_system()`, in order to check if
+  the fiber is not cancellable `fiber:is_system()` can be used now (gh-7473)
```
I wouldn't expose this to public API. The way to make a fiber non-system makes this patch nop, because I can just call `fiber_set_system(false)` on any "system" fiber and then cancel it. And it will work just like without this patchset at all - unexpected problems with internal cancelled fibers.

Gerold103 commented on src/lua/fiber.c:
```diff
@@ -667,13 +667,15 @@ static int
-	 * f == fiber().
-	 */
-	luaL_testcancel(L);
+	if (!(f->flags & FIBER_IS_SYSTEM)) {
+		fiber_cancel(f);
+		/*
+		 * Check if we're ourselves cancelled.
+		 * This also implements cancel for the case when
+		 * f == fiber().
+		 */
+		luaL_testcancel(L);
+	}
```
I still can cancel such fiber via C API. `fiber_cancel()` is public.

Serpentian replied:
> I think that a net.box worker fiber, like any other system fiber, should ignore `fiber.cancel` coming from the Lua world.

If I understood @locker right, we want to limit killing of the fiber only from Lua. From C is should work just the same as it used to do before this patch.

locker replied:
AFAICS it's impossible to look up a system fiber handle in C (`fiber_find` isn't public) so even though `fiber_cancel` is public, one can't possibly cancel a system fiber with it. So I think it's okay to leave `fiber_cancel` as is. Anyway, in future we can update it to ignore system fibers if we need to and add `fiber_cancel_system` internal function for stopping system fibers, but now it doesn't seem to be worthwhile.

Gerold103 replied:
You can reach system fiber, if you make a C function, load it via `box.lib`, and then call it in any netbox trigger. In this C function you can make `fiber_cancel(fiber_self())` and that will kill the system fiber. This entire procedure uses only public API and no FFI hacks.

Totktonada replied:
I agree here with Vladislav: it seems counter-intuitive to have different behavior of C's and Lua's fiber cancelling function, especially when the flag itself is a property of C's fiber.

Alexander Lyapunov recently speak on Arch meeting about C and Lua fiber differences and the general point was that it is not nice to have them. I would not increase the gap without a serious reason.

So I ask for the reason of implementing it this way (only for Lua's fiber cancel). Whether it is possible to reach system fibers from C or not, whether it will be possible in a future or not -- those are not reasons to don't implement it in C, those are reason why it is possible to implement it in Lua. IOW, what is the problem with moving the logic to C's `fiber_cancel()`?

Totktonada replied:
At least, the `FIBER_IS_SYSTEM` name suggests that it has no relation to C or Lua. If it would be `FIBER_IS_CANCELLABLE_FROM_LUA`, it would not raise those questions.

However, it would raise the question what is the difference between C and Lua here? Both are legal languages to write a module, a storage procedure or an application logic.

locker replied:
This PR is actually a step forward - it makes it impossible to cancel a system fiber from normal Lua code, without usage of C modules, thus making Tarantool safer. I agree that we'll probably have to fix C API `fiber_cancel` as well sooner or later, but I see no point in laying off this PR because of this. Let's file a separate ticket for the C API issue and consider fixing it later.

Gerold103 commented on src/box/lua/net_box.c:
```diff
@@ -2717,6 +2717,7 @@ luaT_netbox_transport_start(struct lua_State *L)
 		transport->self_ref = LUA_NOREF;
 		return luaT_error(L);
 	}
```
This `FIBER_IS_SYSTEM` IMO is just an attempt to resurrect `FIBER_IS_CANCELLABLE` (which is not even fully deleted yet). I am not against it and can't propose anything better though (if we want to fix this netbox bug by making the fiber invincible). I only vote against making it a part of the public API - it truly makes the fix useless as people can easily make any fiber non-system again. I wouldn't bother even with `fiber:is_system()` and would indeed just ignore an attempt to cancel a system fiber. I trust Vova to make the further review as one closer to the core team. I can join in the end.

Gerold103 reviewed (commented):
Thanks for the fixes!

locker commented on src/lib/core/fiber.c:
```diff
@@ -557,6 +557,15 @@ fiber_set_joinable(struct fiber *fiber, bool yesno)
 }
 
+void
+fiber_set_system(struct fiber *fiber, bool yesno)
```
I don't think that anybody would ever want to clear this flag or set it after the fiber is stated. I think it's enough to pass it to `fiber_new_ex` in flags. For convenience, I'd add a helper to create a system fiber  - `fiber_new_system`.

locker commented on src/lua/fiber.c:
```diff
@@ -667,13 +667,15 @@ static int
+		 * Check if we're ourselves cancelled.
+		 * This also implements cancel for the case when
+		 * f == fiber().
+		 */
```
Nit: Please always try to keep the diff as small as possible provided, of course, it doesn't hurt code readability - it's good for history. Here, you could apply the following, much less intrusive diff while still keeping the code easy to read:

```diff
diff --git a/src/lua/fiber.c b/src/lua/fiber.c
index ea5bd1a1b209..74b2052641d9 100644
--- a/src/lua/fiber.c
+++ b/src/lua/fiber.c
@@ -667,6 +667,10 @@ static int
 lbox_fiber_cancel(struct lua_State *L)
 {
 	struct fiber *f = lbox_checkfiber(L, 1);
+	if ((f->flags & FIBER_IS_SYSTEM) != 0) {
+		/* System fibers can't be cancelled from Lua. */
+		return 0;
+	}
 	fiber_cancel(f);
 	/*
 	 * Check if we're ourselves cancelled.
```

Serpentian replied:
I wanted to make code in 'happy-path' way. AFAIK, it's supposed to work faster due to processor's optimization mechanisms. But I won't insist and fix it.

locker replied:
This function is definitely not a hot path so I don't think we should bother. Anyway, if you want to explicitly tell the compiler that one branch is executed more often than another, you should use [`likely()`](https://github.com/tarantool/tarantool/blob/7ec71b4f6b0a15997553529c8e15c2dd415df44f/src/trivia/util.h#L159-L170).

locker commented:
> They're a number of fibers, which can be marked as system ones. All of them are not mentioned in cartridge's list. Do we want to do that? IMO this patch seems to be pretty dangerous even with changes, that have been already done.
> 
> * [relay](https://github.com/tarantool/tarantool/blob/969b76ac9901840156c6b9fad45d70ab5de6f221/src/box/relay.cc#L880);

Not necessary. It runs in a different thread (relay).

> * [hot_standby](https://github.com/tarantool/tarantool/blob/969b76ac9901840156c6b9fad45d70ab5de6f221/src/box/recovery.cc#L531);

Yes. Cancelling a hot standby fiber would probably result in a crash or data corruption.

> * [vinyl_task](https://github.com/tarantool/tarantool/blob/969b76ac9901840156c6b9fad45d70ab5de6f221/src/box/vy_scheduler.c#L438-L439);

The link points to "vinyl.scheduler", which must never be cancelled. I think you meant ["task"](https://github.com/tarantool/tarantool/blob/969b76ac9901840156c6b9fad45d70ab5de6f221/src/box/vy_scheduler.c#L1783). We don't need to mark it system, because this fiber runs in a different thread (vinyl worker thread).

> * [box.watcher](https://github.com/tarantool/tarantool/blob/969b76ac9901840156c6b9fad45d70ab5de6f221/src/box/watcher.c#L299-L300);

This fiber runs a user-defined callback so it should be fine to cancel it.

> * [trigger](https://github.com/tarantool/tarantool/blob/969b76ac9901840156c6b9fad45d70ab5de6f221/src/lib/core/trigger.cc#L116);

We use this one to run `box.ctl.on_shutdown` triggers. It should be okay to cancel it.

Gerold103 commented:
Please, make `raft_worker` system fiber too. And grep all the other `fiber_new()` calls in our C code to check every place in case anything else is missed.

locker commented on test/app-luatest/system_fiber_test.lua:
```diff
@@ -0,0 +1,45 @@
+        end
+    end)
+
+    g.server:exec(function()
```
Why do you need the second exec? Can't you put all in one exec, without using `_G`?

Serpentian replied:
You're right. There's no need for `_G` at all. The test looks much cleaner without it.

Serpentian commented:
> And grep all the other `fiber_new()` calls in our C code to check every place in case anything else is missed.

Yep, that's definitely the thing I do. Here's a list of fibers which are not marked as system ones:

* [relay](https://github.com/tarantool/tarantool/blob/969b76ac9901840156c6b9fad45d70ab5de6f221/src/box/relay.cc#L880)
* [vinyl_task](https://github.com/tarantool/tarantool/blob/969b76ac9901840156c6b9fad45d70ab5de6f221/src/box/vy_scheduler.c#L1783)
* [box.watcher](https://github.com/tarantool/tarantool/blob/969b76ac9901840156c6b9fad45d70ab5de6f221/src/box/watcher.c#L299-L300)
* [trigger](https://github.com/tarantool/tarantool/blob/969b76ac9901840156c6b9fad45d70ab5de6f221/src/lib/core/trigger.cc#L116)

(we've already discussed the above mentioned ones)

* [fiber_pool](https://github.com/tarantool/tarantool/blob/969b76ac9901840156c6b9fad45d70ab5de6f221/src/lib/core/fiber_pool.c#L133)
* [main_cord](https://github.com/tarantool/tarantool/blob/969b76ac9901840156c6b9fad45d70ab5de6f221/src/lib/core/fiber.c#L1766)
* [swim_event_handler](https://github.com/tarantool/tarantool/blob/969b76ac9901840156c6b9fad45d70ab5de6f221/src/lib/swim/swim.c#L1923-L1924)
* [on_shutdown_fiber](https://github.com/tarantool/tarantool/blob/969b76ac9901840156c6b9fad45d70ab5de6f221/src/main.cc#L767-L769)
* [lua](https://github.com/tarantool/tarantool/blob/969b76ac9901840156c6b9fad45d70ab5de6f221/src/lua/fiber.c#L403)

> * [main_cord](https://github.com/tarantool/tarantool/blob/969b76ac9901840156c6b9fad45d70ab5de6f221/src/lib/core/fiber.c#L1766)
>  * [lua](https://github.com/tarantool/tarantool/blob/969b76ac9901840156c6b9fad45d70ab5de6f221/src/lua/fiber.c#L403)

It's explicitly said in [cartridge](https://github.com/tarantool/cartridge/blob/d8d48a17e9bf4505941f7f5785fd4b6a9d809918/cartridge/hotreload.lua#L76-L77) that we have to cancel these ones.

>  [fiber_pool](https://github.com/tarantool/tarantool/blob/969b76ac9901840156c6b9fad45d70ab5de6f221/src/lib/core/fiber_pool.c#L133)

Fiber pool is used only [once](https://github.com/tarantool/tarantool/blob/969b76ac9901840156c6b9fad45d70ab5de6f221/src/box/box.cc#L3850-L3852). Should these fibers marked as system? If so, we need to introduce `fiber_pool_create_system` method.

> * [swim_event_handler](https://github.com/tarantool/tarantool/blob/969b76ac9901840156c6b9fad45d70ab5de6f221/src/lib/swim/swim.c#L1923-L1924)

Don't really know about this one. It doesn't look like a system fiber and it should be okay to cancel it too. 

> * [on_shutdown_fiber](https://github.com/tarantool/tarantool/blob/969b76ac9901840156c6b9fad45d70ab5de6f221/src/main.cc#L767-L769)

As we allow to cancel triggers, it should be okay to cancel this fiber too.

---
Cancelling [start-up_script(interactive)](https://github.com/tarantool/tarantool/blob/969b76ac9901840156c6b9fad45d70ab5de6f221/src/lua/init.c#L1002) fiber seems to be pretty critical. I think we should make it system. ?

Serpentian commented on src/lib/core/fiber.h:
```diff
@@ -243,6 +247,21 @@ fiber_self(void);
+ * \sa fiber_start
+ */
+API_EXPORT struct fiber *
+fiber_new_system(const char *name, fiber_func f);
```
Done

Serpentian commented on src/box/applier.cc:
```diff
@@ -173,7 +173,7 @@ applier_fiber_new(struct applier *applier, const char *name, fiber_func func)
 	int pos = snprintf(buf, sizeof(buf), "%s/", name);
 	uri_format(buf + pos, sizeof(buf) - pos, &applier->uri, false);
-	struct fiber *f = fiber_new_xc(buf, func);
+	struct fiber *f = fiber_new_system(buf, func);
```
Done

Gerold103 commented:
> fiber_pool

No, should be good. It seems the pool will just create new fibers instead of the cancelled ones.

> swim_event_handler

Whooops! This one better be protected. Otherwise it can render `swim` object useless (not like it is used by anybody though, lol).

> on_shutdown_fiber

Personally, I would protect this one. It even says in the comment *"Make sure the fiber is non-cancellable"* imposing that it wants to be protected from unintended cancelation.

> start-up_script

Seems fine to cancel it. It will terminate the process most likely, but gracefully. The code is ready to the script fiber eventually ending, it seems. But I would test it manually to be sure.

Serpentian commented:
> Seems fine to cancel it. It will terminate the process most likely, but gracefully.

You're right, it's terminated gracefully. Added swim and on-shutdown, deleted interactive

Totktonada commented:
> NO_DOC=internal
> NO_CHANGELOG=internal

It fixes a user visible problem, how it can be internal?

Serpentian commented:
Rebased, fixed commit message and added changelog.

Gerold103 commented on test/app-luatest/system_fiber_test.lua:
```diff
@@ -0,0 +1,41 @@
+            end
+        end
+
+        fiber.yield()
+        t.assert(conn:ping())
+        for _, f in pairs(system_fibers) do
+            t.assert(f:status() ~= 'dead')
```
It is not really enough to protect only the Lua land. As I was saying, I can get to almost any system fiber via C API. And I don't need `fiber.find()` for this. System fibers often execute triggers where I can get `fiber_self()` and do with it anything:
```Lua
lib = box.lib.load('test')
kill_self = lib:load('oh_yes_kill_me')

net = require('net.box')
box.cfg{listen = 3313}
box.schema.user.grant('guest', 'super')
c = net.connect(box.cfg.listen, {reconnect_after = 1})
c:ping()

c:eval([[box.session.push(nil)]], {}, {on_push = function()
    kill_self()
end})
```
```C
#include "module.h"

int
oh_yes_kill_me(box_function_ctx_t *ctx, const char *args, const char *args_end)
{
	(void)ctx, (void)args, (void)args_end;
	fiber_cancel(fiber_self());
	return 0;
}
```
Compilation on Mac. `test.c` is stored in `tarantool/sandbox` with the C code above.
```
clang -dynamiclib -undefined dynamic_lookup -I../src/ -I../third_party/luajit/src/ -I../src/lib/msgpuck/ test.c -o test.dylib
```
Then copy the `test.dylib` into tarantool's workdir. This example is a bit exaggerated, but killing usually happens accidentally via more creative logic than here.

Serpentian replied:
We cannot delete `fiber_cancel` from the fiber's public C API. But can we change its API?

If so we can make `fiber_cancel` a wrapper around the function, the content of which will be equivalent to the current `fiber_cancel` (don't really know how to name it)). In this wrapper we can check if the fiber is system. If so we will set `diag` and return -1. We won't export internal function to public API. So, the fiber won't be ever cancelled from public. In tarantool itself we gonna use only this internal function. But we need to change the return value of `fiber_cancel`.  

@Gerold103 what do you think about this?

locker replied:
Since nobody complained so far, I'd ignore this issue for now. We can always fix it later if we need to.

Serpentian replied:
Considering the fact, that the release is pretty close, we should merge something anyway. Overwise, there will be one more release with `error_reconnect` bug, which is actually solved by this PR. 

Now the fiber can be cancelled with sophisticated workarounds.  The possible solution: 

Let's make public C API function `fiber_cancel()` a wrapper, which will check if the fiber is system:
```C
void
fiber_cancel(struct fiber *f)
{
	/* Do nothing if the FIBER_IS_SYSTEM flag is set. */
	if ((f->flags & FIBER_IS_SYSTEM) != 0)
		return;

	fiber_cancel_system(f);
}
```
The content of the `fiber_cancel_system` is the same as current `fiber_cancel`.

We cannot change the API of the `fiber_cancel`, so let's just silently ignore cancellation request for a system fiber. This behavior is similar to the current one, but all checks will be done in public C API, and both API will behave the same way, as @Totktonada proposed. Moreover, it will be impossible to cancel the fiber from anywhere, as @Gerold103 proposed.

I created another PR for that solution: #7524. If it's all right or some minor changes are needed we can merge it. Overwise, 
let's merge this one, create ticket for cancelling system fibers from C and I will fix it ASAP after release.

Totktonada replied:
`fiber_cancel()` always ignores system fibers, we just had no system fibers before. If I'll say it this way, it will not be incompatible change.

If you're bother about API breakage, I completely missed point of this PR. Why this 'break' of C API is not okay, but the same is okay for the Lua API, which is more frequently used?

I requested a reason of the difference between C's and Lua's fiber cancel behavior (in a discussion above, not answered) and if your answer is 'backward compatibility', the answer is wrong. Technically speaking, cancelling of fibers that goes to be marked as system ones never produces a useful result. See compatibility terms definition in the [new release policy](https://github.com/tarantool/tarantool/discussions/6182) document:

> A code that are run on an older release is run with the same effect on a
newer release. However only meaningful code counts: if any code just give an
error, but starts doing something useful, the change is considered
compatible.

If you're bother about error reporting in `fiber_cancel()` (return value, box error area), just don't report an error and silently do nothing for a system fiber. We can add `fiber_cancel_v2()` later if there will be a demand.

Gerold103 replied:
* `fiber_cancel()` always ignored "system" fibers. Except that the "systemness" was defined per-scope and using `FIBER_IS_CANCELLABLE` which we are trying to resurrect here in a better form;
* We don't need wrappers around it. Just make it not touch the system fibers. For forceful internal killing we need a new function with a good name (not `fiber_cancel_internal()`). Maybe `fiber_kill_sys()` or just `fiber_kill()` (a bit conflicting with Lua) or `fiber_cancel_forced()` or likewise. It shouldn't be exported to the public API in any way.
* About one more release with `error_reconnect` bug - I see no problem with that. The moment it appeared in at least one release we've already got it public. Doesn't matter now in how many releases. At least in version-flexible modules like vshard - they will need to workaround this bug anyway, whether you fix it now or next time.

As I was saying, I can't see much harm in the current form of this PR, but can't see much usefulness too. I can't vote it up, but Vova's ack should be enough anyway if you decide to move forward with this partial fix.

locker replied:
`fiber_cancel_system` (matches `fiber_new_system`) sounds fine by me, but I'd be fine with doing this some time later, in a separate PR, as I said above. Anyway, I can't merge this PR unitl @Gerold103 approves:

<details>

![image](https://user-images.githubusercontent.com/1114511/182805926-431e7cc6-5791-4118-958f-0a9edcaeb709.png)

</details>

Serpentian replied:
Let's not merge this PR before the full solution is done. Current one is not full and the behavior of Lua's and C's `fiber_cancel` differ, which is not too great. 

----

> fiber_cancel() always ignored "system" fibers. Except that the "systemness" was defined per-scope and using FIBER_IS_CANCELLABLE which we are trying to resurrect here in a better form;

Didn't get your statement. `FIBER_IS_CANCELLABLE` is used only in order not to wakeup the fiber after its cancelling. However, currently all fibers can be cancelled if they call `fiber_testcancel()` inside. AFAICS, we don't want to set the flag `FIBER_IS_CANCELLED` in case of system fiber at all. 

----

> We don't need wrappers around it.

Why not? The logic of the `fiber_cancel` should remain the same, excepting checking for systemness. `fiber_cancel_system` is supposed to work the same way the `fiber_cancel` currently works. If I'm not right, could you please explain, why and what will be the differences in behavior of the current `fiber_cancel` and `fiber_cancel_system`. 

----

> About one more release with error_reconnect bug - I see no problem with that.

Got you, lets postpone the merging.

Gerold103 replied:
> Didn't get your statement. FIBER_IS_CANCELLABLE is used only in order not to wakeup the fiber after its cancelling

В теории это так. По факту его использовали где ни попадя как признак "системности".

> The logic of the fiber_cancel should remain the same, excepting checking for systemness.

`fiber_cancel()` можно менять. Его можно заставить игнорить системные файберы, в этом нет проблемы.

> fiber_cancel_system is supposed to work the same way the fiber_cancel currently works

Yes.

> what will be the differences in behavior

`fiber_cancel()` ignores system fibers and it is public, `fiber_cancel_system()` doesn't ignore the system fibers and is private (not in module.h, not in exports file).

In the meantime I found another related bug: https://github.com/tarantool/tarantool/issues/7537. We probably need a more serious solution which would cover all those recent fiber issues.

Serpentian replied:
> fiber_cancel() ignores system fibers and it is public, fiber_cancel_system() doesn't ignore the system fibers and is private

Exactly the thing which is done here: https://github.com/tarantool/tarantool/pull/7524. However, `fiber_cancel_system` is added to fiber.h as there's no internal fiber API and, of course, it doesn't cover #7537.

Gerold103 replied:
The problem is not in `fiber.h`. Only one part of `fiber.h` is public. See the comments `/** \cond public */` and `/** \endcond public */` in the file.

Gerold103 replied:
> Anyway, I can't merge this PR unitl @Gerold103 approves

Just tell me and I will approve 😉

locker replied:
> > Anyway, I can't merge this PR unitl @Gerold103 approves
> 
> Just tell me and I will approve wink

@Gerold103 I think that this ticket does resolve #7473 - it actually prohibits killing system fibers via Lua API. True, we can still kill a system fiber via C API, but I think we should file separate tickets for this and merge this PR as is, because it's a step forward. Please approve.

Gerold103 reviewed (commented):
Thanks for the patch! While I can't see major problems, neither can I see the bug being gone - still can kill system fibers using just public API and no FFI. I remember I already mentioned how to do it earlier, but was told it is fine. In this case you can ignore the last comment and proceed with pushing to master. Then I will file it as a separate ticket.

Totktonada commented on src/lib/core/fiber.h:
```diff
@@ -170,6 +170,10 @@ enum {
 	 * This flag is set when fiber has custom max slice.
 	 */
 	FIBER_CUSTOM_SLICE	= 1 << 8,
+	/**
+	 * This flag idicates, if the fiber can be killed from the Lua world.
+	 */
+	FIBER_IS_SYSTEM         = 1 << 9,
```
Just to bring some attention of people who are working on it: I placed some thoughts around this approach to fix cartridge's weird hot reload into the issue: https://github.com/tarantool/tarantool/issues/7473#issuecomment-1203235216.

Totktonada replied:
And... whether anybody consider dropping the code that cancels fibers from cartridge itself? Why we can't?

Serpentian commented on src/lib/core/fiber.h:
```diff
@@ -1094,6 +1112,20 @@ fiber_new_xc(const char *name, fiber_func func)
 	return f;
 }
 
+/*
```
Fixed

Serpentian commented on src/main.cc:
```diff
@@ -29,6 +29,7 @@
  * SUCH DAMAGE.
  */
 #include "main.h"
+#include "lib/core/fiber.h"
```
Fixed. This is some kind of emacs extension which does it automatically, the one, which I cannot find and turn off. It wasn't on purpose

Gerold103 reviewed (approved):
As you wish.

locker commented:
Cherry-picked to 2.10.
